proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1205 @@
+      # each arm of the variant
+      payload_types = flatten_cases(direction, $cases)
+      for i in 0..result.len() - 1:
+        result[i + 1] = upcast(result[i + 1], payload_types[i])
+
+      # If the shared type has more values than this payload generated we
+      # insert zero values to pad
+      for i in result.len() - 1..payload_types.len()
+        result.push(zero_value(payload_types[i]))
+
+      result
+    }
+
+    # Lowering a resource means that we're inserting our handle into a fresh
+    # slot in the destination's table. This insertion operation allocates a new
+    # i32 value which will point to the handle, and note that this also bumps the
+    # handle's internal reference count.
+    handle<$resource>(val) => [TABLES.insert($resource, val, dst)],
+
+    # Buffers are only ever lowered as parameters of an exported function.
+    # Buffers can't be returned from a function so they can never be lowered
+    # for import results.
+    push-buffer<$ty>(val) | pull-buffer<$ty>(val) => {
+      assert(direction == "export")
+      [TABLES.insert_buffer(val, dst)]
+    }
+
+def lower_to_memory(direction, dst, ptr, gen):
+  match gen.next()
+    f32(val) => write_4_bytes(dst, ptr, val),
+    f64(val) => write_8_bytes(dst, ptr, val),
+
+    s8(val) => write_1_byte(dst, ptr, val),
+    u8(val) => write_1_byte(dst, ptr, val),
+    s16(val) => write_2_bytes(dst, ptr, val),
+    u16(val) => write_2_bytes(dst, ptr, val),
+    s32(val) => write_4_bytes(dst, ptr, val),
+    u32(val) => write_4_bytes(dst, ptr, val),
+    s64(val) => write_8_bytes(dst, ptr, val),
+    u64(val) => write_8_bytes(dst, ptr, val),
+
+    char(val) => write_4_bytes(dst, ptr, val),
+
+    string(val) |
+    list<$ty>(val) => {
+      gen.requeue(val) # put `val` back in the generator to get yielded first in `lower`
+      [val_ptr, val_len] = lower(direction, dst, val)
+      write_4_bytes(dst, ptr, val_ptr)
+      write_4_bytes(dst, ptr + 4, val_len)
+    }
+
+    flags<$fields>(val) => {
+      gen.requeue(val)
+      for i, bits in enumerate(lower(direction, dst, val)):
+        write_4_bytes(dst, ptr + i * 4, bits)
+    }
+
+    record_start<$fields> => {
+      offset = 0
+      for $field in $fields:
+        offset = align_to(offset, align($field))
+        lower_to_memory(direction, dst, ptr + offset, gen)
+        offset += size(direction, $field)
+      assert_eq(gen.next(), record_end<$fields>) # true by construction
+    }
+
+    variant_start<$cases>(label) => {
+    discrim = $cases.index(label)
+      offset = 0
+      match variant_discriminant($cases) {
+        i8 => {
+          write_1_byte(dst, ptr, discrim)
+          offset += 1
+        }
+        i16 => {
+          write_2_bytes(dst, ptr, discrim)
+          offset += 2
+        }
+        i32 => {
+          write_4_bytes(dst, ptr, discrim)
+          offset += 4
+        }
+        i64 => {
+          write_8_bytes(dst, ptr, discrim)
```

Comment:
Hm, here and elsewhere, we are handling i64 discrimints, but we always lower discriminats as wasm_i32, so perhaps this should just be unreachable?